File: ts/src/NewtonSDK.ts
// Newton Ts SDK

import { AbsEntity } from './entity/AbsEntity'
import { ArccoEntity } from './entity/ArccoEntity'
import { ArcsinEntity } from './entity/ArcsinEntity'
import { ArctanEntity } from './entity/ArctanEntity'
import { AreaEntity } from './entity/AreaEntity'
import { CosEntity } from './entity/CosEntity'
import { DeriveEntity } from './entity/DeriveEntity'
import { FactorEntity } from './entity/FactorEntity'
import { IntegrateEntity } from './entity/IntegrateEntity'
import { LogEntity } from './entity/LogEntity'
import { SimplifyEntity } from './entity/SimplifyEntity'
import { SinEntity } from './entity/SinEntity'
import { TanEntity } from './entity/TanEntity'
import { TangentEntity } from './entity/TangentEntity'
import { ZeroEntity } from './entity/ZeroEntity'

export type * from './NewtonTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { NewtonEntityBase } from './NewtonEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class NewtonSDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _abs?: AbsEntity

  // Idiomatic facade: `client.abs.list()` / `client.abs.load({ id })`.
  get abs(): AbsEntity {
    return (this._abs ??= new AbsEntity(this, undefined))
  }

  /** @deprecated Use `client.abs` instead. */
  Abs(data?: any) {
    const self = this
    return new AbsEntity(self,data)
  }


  _arcco?: ArccoEntity

  // Idiomatic facade: `client.arcco.list()` / `client.arcco.load({ id })`.
  get arcco(): ArccoEntity {
    return (this._arcco ??= new ArccoEntity(this, undefined))
  }

  /** @deprecated Use `client.arcco` instead. */
  Arcco(data?: any) {
    const self = this
    return new ArccoEntity(self,data)
  }


  _arcsin?: ArcsinEntity

  // Idiomatic facade: `client.arcsin.list()` / `client.arcsin.load({ id })`.
  get arcsin(): ArcsinEntity {
    return (this._arcsin ??= new ArcsinEntity(this, undefined))
  }

  /** @deprecated Use `client.arcsin` instead. */
  Arcsin(data?: any) {
    const self = this
    return new ArcsinEntity(self,data)
  }


  _arctan?: ArctanEntity

  // Idiomatic facade: `client.arctan.list()` / `client.arctan.load({ id })`.
  get arctan(): ArctanEntity {
    return (this._arctan ??= new ArctanEntity(this, undefined))
  }

  /** @deprecated Use `client.arctan` instead. */
  Arctan(data?: any) {
    const self = this
    return new ArctanEntity(self,data)
  }


  _area?: AreaEntity

  // Idiomatic facade: `client.area.list()` / `client.area.load({ id })`.
  get area(): AreaEntity {
    return (this._area ??= new AreaEntity(this, undefined))
  }

  /** @deprecated Use `client.area` instead. */
  Area(data?: any) {
    const self = this
    return new AreaEntity(self,data)
  }


  _cos?: CosEntity

  // Idiomatic facade: `client.cos.list()` / `client.cos.load({ id })`.
  get cos(): CosEntity {
    return (this._cos ??= new CosEntity(this, undefined))
  }

  /** @deprecated Use `client.cos` instead. */
  Cos(data?: any) {
    const self = this
    return new CosEntity(self,data)
  }


  _derive?: DeriveEntity

  // Idiomatic facade: `client.derive.list()` / `client.derive.load({ id })`.
  get derive(): DeriveEntity {
    return (this._derive ??= new DeriveEntity(this, undefined))
  }

  /** @deprecated Use `client.derive` instead. */
  Derive(data?: any) {
    const self = this
    return new DeriveEntity(self,data)
  }


  _factor?: FactorEntity

  // Idiomatic facade: `client.factor.list()` / `client.factor.load({ id })`.
  get factor(): FactorEntity {
    return (this._factor ??= new FactorEntity(this, undefined))
  }

  /** @deprecated Use `client.factor` instead. */
  Factor(data?: any) {
    const self = this
    return new FactorEntity(self,data)
  }


  _integrate?: IntegrateEntity

  // Idiomatic facade: `client.integrate.list()` / `client.integrate.load({ id })`.
  get integrate(): IntegrateEntity {
    return (this._integrate ??= new IntegrateEntity(this, undefined))
  }

  /** @deprecated Use `client.integrate` instead. */
  Integrate(data?: any) {
    const self = this
    return new IntegrateEntity(self,data)
  }


  _log?: LogEntity

  // Idiomatic facade: `client.log.list()` / `client.log.load({ id })`.
  get log(): LogEntity {
    return (this._log ??= new LogEntity(this, undefined))
  }

  /** @deprecated Use `client.log` instead. */
  Log(data?: any) {
    const self = this
    return new LogEntity(self,data)
  }


  _simplify?: SimplifyEntity

  // Idiomatic facade: `client.simplify.list()` / `client.simplify.load({ id })`.
  get simplify(): SimplifyEntity {
    return (this._simplify ??= new SimplifyEntity(this, undefined))
  }

  /** @deprecated Use `client.simplify` instead. */
  Simplify(data?: any) {
    const self = this
    return new SimplifyEntity(self,data)
  }


  _sin?: SinEntity

  // Idiomatic facade: `client.sin.list()` / `client.sin.load({ id })`.
  get sin(): SinEntity {
    return (this._sin ??= new SinEntity(this, undefined))
  }

  /** @deprecated Use `client.sin` instead. */
  Sin(data?: any) {
    const self = this
    return new SinEntity(self,data)
  }


  _tan?: TanEntity

  // Idiomatic facade: `client.tan.list()` / `client.tan.load({ id })`.
  get tan(): TanEntity {
    return (this._tan ??= new TanEntity(this, undefined))
  }

  /** @deprecated Use `client.tan` instead. */
  Tan(data?: any) {
    const self = this
    return new TanEntity(self,data)
  }


  _tangent?: TangentEntity

  // Idiomatic facade: `client.tangent.list()` / `client.tangent.load({ id })`.
  get tangent(): TangentEntity {
    return (this._tangent ??= new TangentEntity(this, undefined))
  }

  /** @deprecated Use `client.tangent` instead. */
  Tangent(data?: any) {
    const self = this
    return new TangentEntity(self,data)
  }


  _zero?: ZeroEntity

  // Idiomatic facade: `client.zero.list()` / `client.zero.load({ id })`.
  get zero(): ZeroEntity {
    return (this._zero ??= new ZeroEntity(this, undefined))
  }

  /** @deprecated Use `client.zero` instead. */
  Zero(data?: any) {
    const self = this
    return new ZeroEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new NewtonSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return NewtonSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'Newton' }
  }

  toString() {
    return 'Newton ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = NewtonSDK


export {
  stdutil,

  BaseFeature,
  NewtonEntityBase,

  NewtonSDK,
  SDK,
}

File: go/core/newton_sdk.go
package core

import (
	"fmt"

	vs "github.com/voxgig-sdk/newton-sdk/go/utility/struct"
)

type NewtonSDK struct {
	Mode     string
	options  map[string]any
	utility  *Utility
	Features []Feature
	rootctx  *Context
}

func NewNewtonSDK(options map[string]any) *NewtonSDK {
	sdk := &NewtonSDK{
		Mode:     "live",
		Features: []Feature{},
	}

	sdk.utility = NewUtility()

	config := MakeConfig()

	sdk.rootctx = sdk.utility.MakeContext(map[string]any{
		"client":  sdk,
		"utility": sdk.utility,
		"config":  config,
		"options": options,
		"shared":  map[string]any{},
	}, nil)

	sdk.options = sdk.utility.MakeOptions(sdk.rootctx)

	if vs.GetPath([]any{"feature", "test", "active"}, sdk.options) == true {
		sdk.Mode = "test"
	}

	sdk.rootctx.Options = sdk.options

	// Add features in the resolved order (MakeOptions puts an explicit array
	// order first, else defaults to test-first). Ordering matters: the `test`
	// feature installs the base mock transport and the transport features
	// (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
	// must be added before them to sit at the base of the chain.
	featureOpts := ToMapAny(vs.GetProp(sdk.options, "feature"))
	if featureOpts != nil {
		if fo, ok := vs.GetPath([]any{"__derived__", "featureorder"}, sdk.options).([]any); ok {
			for _, n := range fo {
				fname, _ := n.(string)
				fopts := ToMapAny(featureOpts[fname])
				if fopts != nil {
					if active, ok := fopts["active"]; ok {
						if ab, ok := active.(bool); ok && ab {
							sdk.utility.FeatureAdd(sdk.rootctx, makeFeature(fname))
						}
					}
				}
			}
		}
	}

	// Add extension features.
	if extend := vs.GetProp(sdk.options, "extend"); extend != nil {
		if extList, ok := extend.([]any); ok {
			for _, f := range extList {
				if feat, ok := f.(Feature); ok {
					sdk.utility.FeatureAdd(sdk.rootctx, feat)
				}
			}
		}
	}

	// Initialize features.
	for _, f := range sdk.Features {
		sdk.utility.FeatureInit(sdk.rootctx, f)
	}

	sdk.utility.FeatureHook(sdk.rootctx, "PostConstruct")

	return sdk
}

func (sdk *NewtonSDK) OptionsMap() map[string]any {
	out := vs.Clone(sdk.options)
	if om, ok := out.(map[string]any); ok {
		return om
	}
	return map[string]any{}
}

func (sdk *NewtonSDK) GetUtility() *Utility {
	return CopyUtility(sdk.utility)
}

func (sdk *NewtonSDK) GetRootCtx() *Context {
	return sdk.rootctx
}

func (sdk *NewtonSDK) Prepare(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "prepare",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	options := sdk.options

	path, _ := vs.GetProp(fetchargs, "path").(string)
	method, _ := vs.GetProp(fetchargs, "method").(string)
	if method == "" {
		method = "GET"
	}

	params := ToMapAny(vs.GetProp(fetchargs, "params"))
	if params == nil {
		params = map[string]any{}
	}
	query := ToMapAny(vs.GetProp(fetchargs, "query"))
	if query == nil {
		query = map[string]any{}
	}

	headers := utility.PrepareHeaders(ctx)

	base, _ := vs.GetProp(options, "base").(string)
	prefix, _ := vs.GetProp(options, "prefix").(string)
	suffix, _ := vs.GetProp(options, "suffix").(string)

	ctx.Spec = NewSpec(map[string]any{
		"base":    base,
		"prefix":  prefix,
		"suffix":  suffix,
		"path":    path,
		"method":  method,
		"params":  params,
		"query":   query,
		"headers": headers,
		"body":    vs.GetProp(fetchargs, "body"),
		"step":    "start",
	})

	// Merge user-provided headers.
	if uh := vs.GetProp(fetchargs, "headers"); uh != nil {
		if uhm, ok := uh.(map[string]any); ok {
			for k, v := range uhm {
				ctx.Spec.Headers[k] = v
			}
		}
	}

	_, err := utility.PrepareAuth(ctx)
	if err != nil {
		return nil, err
	}

	return utility.MakeFetchDef(ctx)
}

func (sdk *NewtonSDK) Direct(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	fetchdef, err := sdk.Prepare(fetchargs)
	if err != nil {
		return map[string]any{"ok": false, "err": err}, nil
	}

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "direct",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	url, _ := fetchdef["url"].(string)
	fetched, fetchErr := utility.Fetcher(ctx, url, fetchdef)

	if fetchErr != nil {
		return map[string]any{"ok": false, "err": fetchErr}, nil
	}

	if fetched == nil {
		return map[string]any{
			"ok":  false,
			"err": ctx.MakeError("direct_no_response", "response: undefined"),
		}, nil
	}

	if fm, ok := fetched.(map[string]any); ok {
		status := ToInt(vs.GetProp(fm, "status"))
		headers := vs.GetProp(fm, "headers")

		// No-body responses (204, 304) and explicit zero content-length
		// must skip JSON parsing — calling json() on an empty body errors.
		var contentLength string
		if hm, ok := headers.(map[string]any); ok {
			if cl, ok := hm["content-length"]; ok {
				contentLength = fmt.Sprintf("%v", cl)
			}
		}
		noBody := status == 204 || status == 304 || contentLength == "0"

		var jsonData any
		if !noBody {
			if jf := vs.GetProp(fm, "json"); jf != nil {
				if f, ok := jf.(func() any); ok {
					// f() returns nil on parse error in our fetcher.
					jsonData = f()
				}
			}
		}

		return map[string]any{
			"ok":      status >= 200 && status < 300,
			"status":  status,
			"headers": headers,
			"data":    jsonData,
		}, nil
	}

	return map[string]any{"ok": false, "err": ctx.MakeError("direct_invalid", "invalid response type")}, nil
}


// Abs returns a Abs entity bound to this client.
// Idiomatic usage: client.Abs(nil).List(nil, nil) or
// client.Abs(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Abs(data map[string]any) NewtonEntity {
	return NewAbsEntityFunc(sdk, data)
}


// Arcco returns a Arcco entity bound to this client.
// Idiomatic usage: client.Arcco(nil).List(nil, nil) or
// client.Arcco(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Arcco(data map[string]any) NewtonEntity {
	return NewArccoEntityFunc(sdk, data)
}


// Arcsin returns a Arcsin entity bound to this client.
// Idiomatic usage: client.Arcsin(nil).List(nil, nil) or
// client.Arcsin(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Arcsin(data map[string]any) NewtonEntity {
	return NewArcsinEntityFunc(sdk, data)
}


// Arctan returns a Arctan entity bound to this client.
// Idiomatic usage: client.Arctan(nil).List(nil, nil) or
// client.Arctan(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Arctan(data map[string]any) NewtonEntity {
	return NewArctanEntityFunc(sdk, data)
}


// Area returns a Area entity bound to this client.
// Idiomatic usage: client.Area(nil).List(nil, nil) or
// client.Area(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Area(data map[string]any) NewtonEntity {
	return NewAreaEntityFunc(sdk, data)
}


// Cos returns a Cos entity bound to this client.
// Idiomatic usage: client.Cos(nil).List(nil, nil) or
// client.Cos(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Cos(data map[string]any) NewtonEntity {
	return NewCosEntityFunc(sdk, data)
}


// Derive returns a Derive entity bound to this client.
// Idiomatic usage: client.Derive(nil).List(nil, nil) or
// client.Derive(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Derive(data map[string]any) NewtonEntity {
	return NewDeriveEntityFunc(sdk, data)
}


// Factor returns a Factor entity bound to this client.
// Idiomatic usage: client.Factor(nil).List(nil, nil) or
// client.Factor(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Factor(data map[string]any) NewtonEntity {
	return NewFactorEntityFunc(sdk, data)
}


// Integrate returns a Integrate entity bound to this client.
// Idiomatic usage: client.Integrate(nil).List(nil, nil) or
// client.Integrate(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Integrate(data map[string]any) NewtonEntity {
	return NewIntegrateEntityFunc(sdk, data)
}


// Log returns a Log entity bound to this client.
// Idiomatic usage: client.Log(nil).List(nil, nil) or
// client.Log(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Log(data map[string]any) NewtonEntity {
	return NewLogEntityFunc(sdk, data)
}


// Simplify returns a Simplify entity bound to this client.
// Idiomatic usage: client.Simplify(nil).List(nil, nil) or
// client.Simplify(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Simplify(data map[string]any) NewtonEntity {
	return NewSimplifyEntityFunc(sdk, data)
}


// Sin returns a Sin entity bound to this client.
// Idiomatic usage: client.Sin(nil).List(nil, nil) or
// client.Sin(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Sin(data map[string]any) NewtonEntity {
	return NewSinEntityFunc(sdk, data)
}


// Tan returns a Tan entity bound to this client.
// Idiomatic usage: client.Tan(nil).List(nil, nil) or
// client.Tan(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Tan(data map[string]any) NewtonEntity {
	return NewTanEntityFunc(sdk, data)
}


// Tangent returns a Tangent entity bound to this client.
// Idiomatic usage: client.Tangent(nil).List(nil, nil) or
// client.Tangent(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Tangent(data map[string]any) NewtonEntity {
	return NewTangentEntityFunc(sdk, data)
}


// Zero returns a Zero entity bound to this client.
// Idiomatic usage: client.Zero(nil).List(nil, nil) or
// client.Zero(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NewtonSDK) Zero(data map[string]any) NewtonEntity {
	return NewZeroEntityFunc(sdk, data)
}



func TestSDK(testopts map[string]any, sdkopts map[string]any) *NewtonSDK {
	if sdkopts == nil {
		sdkopts = map[string]any{}
	}
	sdkopts = vs.Clone(sdkopts).(map[string]any)

	if testopts == nil {
		testopts = map[string]any{}
	}
	testopts = vs.Clone(testopts).(map[string]any)
	testopts["active"] = true

	vs.SetPath(sdkopts, []any{"feature", "test"}, testopts)

	sdk := NewNewtonSDK(sdkopts)
	sdk.Mode = "test"

	return sdk
}
